extensions/src/main/java/org/pentaho/platform/plugin/services/security/userrole/oauth/PentahoClientRegistrationFactory.java

```diff
@@ -0,0 +1,115 @@
+
+  private Map<String, String> propertiesMap;
+  private Map<String, Object> configurationMetadata;
+
+  public PentahoClientRegistrationFactory( Map<String, String> propertiesMap ) {
+    this.propertiesMap = propertiesMap;
+  }
+
+  public PentahoClientRegistrationFactory( Map<String, String> propertiesMap,
+                                           Map<String, Object> configurationMetadata ) {
+    this.propertiesMap = propertiesMap;
+    this.configurationMetadata = configurationMetadata;
+  }
+
+  @Override
+  public ClientRegistration getObject() {
+    ClientRegistration.Builder clientRegistrationBuilder =
+      ClientRegistration.withRegistrationId( propertiesMap.get( "registrationId" ) );
+
+    if ( propertiesMap.containsKey( "clientId" ) ) {
+      clientRegistrationBuilder.clientId( propertiesMap.get( "clientId" ) );
+    }
+    if ( propertiesMap.containsKey( "clientSecret" ) ) {
+      clientRegistrationBuilder.clientSecret( Encr.decryptPassword( propertiesMap.get( "clientSecret" ) ) );
+    }
+    if ( propertiesMap.containsKey( "authorizationUri" ) ) {
+      clientRegistrationBuilder.authorizationUri( propertiesMap.get( "authorizationUri" ) );
+    }
+    if ( propertiesMap.containsKey( "jwkSetUri" ) ) {
+      clientRegistrationBuilder.jwkSetUri( propertiesMap.get( "jwkSetUri" ) );
+    }
+    if ( propertiesMap.containsKey( "tokenUri" ) ) {
+      clientRegistrationBuilder.tokenUri( propertiesMap.get( "tokenUri" ) );
+    }
+    if ( propertiesMap.containsKey( "userInfoUri" ) ) {
+      clientRegistrationBuilder.userInfoUri( propertiesMap.get( "userInfoUri" ) );
+    }
+    if ( propertiesMap.containsKey( "clientName" ) ) {
+      clientRegistrationBuilder.clientName( propertiesMap.get( "clientName" ) );
+    }
+    if ( propertiesMap.containsKey( "scope" ) ) {
+      clientRegistrationBuilder.scope(
+        Objects.toString( propertiesMap.get( "scope" ), StringUtils.EMPTY ).split( "," ) );
+    }
+    if ( propertiesMap.containsKey( "redirectUri" ) ) {
+      clientRegistrationBuilder.redirectUri( propertiesMap.get( "redirectUri" ) );
+    }
+    if ( propertiesMap.containsKey( "authorizationGrantType" ) ) {
+      clientRegistrationBuilder.authorizationGrantType(
+        new AuthorizationGrantType( propertiesMap.get( "authorizationGrantType" ) ) );
+    }
+    if ( propertiesMap.containsKey( "clientAuthenticationMethod" ) ) {
+      clientRegistrationBuilder.clientAuthenticationMethod(
+        new ClientAuthenticationMethod( propertiesMap.get( "clientAuthenticationMethod" ) ) );
+    }
+    if ( propertiesMap.containsKey( "userInfoAuthenticationMethod" ) ) {
+      clientRegistrationBuilder.userInfoAuthenticationMethod(
+        new AuthenticationMethod( propertiesMap.get( "userInfoAuthenticationMethod" ) ) );
+    }
+    if ( MapUtils.isNotEmpty( this.configurationMetadata ) ) {
+      clientRegistrationBuilder.providerConfigurationMetadata( this.configurationMetadata );
```

Comment:
I don't see a test that verifies all the values are set in the returned ClientRegistration from `PentahoClientRegistrationFactory.java#getObject`